ideas/apq.h

```diff
@@ -0,0 +1,285 @@
+		}
+		auto op = detail::make_request_cursor_op(conn, q, c, h);
+		op.perform();
+	});
+
+	return init.result.get();
+}
+
+template <parameters_transfer_mode prm>
+template <typename T, typename CompletionToken>
+inline auto basic_protocol<prm>::fetch(cursor<T>& c, T& t, CompletionToken token) {
+	detail::async_result_init<CompletionToken, void (error_code, size_type)> init(std::move(token));
+
+	if (!c) {
+		asio_handler_invoke([h = init.handler]{ h(error_code{}, 0);}, std::addressof(h));
+	}
+	auto op = detail::make_fetch_op(c, init.handler);
+	op.perform();
+
+	return init.result.get();
+}
+
+template <parameters_transfer_mode prm>
+template <typename ConnectionProvider, typename CompletionToken>
+inline auto basic_protocol<prm>::update(ConnectionProvider provider, query q, CompletionToken token) {
+	detail::async_result_init<CompletionToken, void (error_code, size_type)> init(std::move(token));
+
+	get_connection(provider, [q, h = init.handler] (error_code ec, connection conn) {
+		if (ec) {
+			return asio_handler_invoke([ec, h]{ h(ec);}, std::addressof(h));
+		}
+		auto op = detail::make_update_op(conn, q, h);
+		op.perform();
+	});
+
+	return init.result.get();
+}
+
+template <parameters_transfer_mode prm>
+template <typename ConnectionProvider, typename CompletionToken>
+inline auto basic_protocol<prm>::execute(ConnectionProvider provider, query q, CompletionToken token) {
+	detail::async_result_init<CompletionToken, void (error_code)> init(std::move(token));
+
+	get_connection(provider, [q, h = init.handler] (error_code ec, connection conn) {
+		if (ec) {
+			return asio_handler_invoke([ec, h]{ h(ec);}, std::addressof(h));
+		}
+		auto op = detail::make_execute_op(conn, q, h);
+		op.perform();
+	});
+
+	return init.result.get();
+}
+
+template <parameters_transfer_mode prm>
+template <typename ConnectionProvider, typename CompletionToken>
+inline auto basic_protocol<prm>::begin(ConnectionProvider provider, CompletionToken token) {
+
+	const auto fwd = [](auto h, error_code ec, transaction_connection c) {
+		asio_handler_invoke([&]{ h(ec, c);}, std::addressof(h)); // io.post()?
+	};
+
+	detail::async_result_init<CompletionToken, void (error_code)> init(std::move(token));
+	
+	get_connection(provider, [fwd, token = init.handler](error_code ec, connection conn) {
+		if (ec) {
+			return fwd(std::move(token), ec, {});
+		}
+		basic_protocol::execute(conn, queries::begin_transaction, 
+				[fwd, conn, token = std::move(token)] (error_code ec) {
+			if (ec) {
+				return fwd(std::move(token), ec, {});
+			}
+			fwd(std::move(token), {}, {conn});
+		});
+	});
+
+	return init.result.get();
+}; // Begin transaction
+
+
+make_query( "SELECT name FROM my_table WHERE id="_sql << id << " AND sid=" << sid );
+
+make_query( "SELECT name FROM my_table WHERE id=:id  AND sid=:sid"_sql % (("id",id), ("sid",sid)) );
+
+
+BOOST_FUSION_DEFINE_STRUCT((), (my_query_params), 
+	(std::string, id)
+	(int, sid)
+);
```

Comment:
We need some type definition mechanism to allow map C++ types to database types through the OIDs and type names. So we need some traits system with easy macro for custom types definition.  We can predefine all the traits for geninue types like int, string and so on. E.g. for composite types we need something like BOOST_FUSION_DEFINE_STRUCT but with type traits definition inside, let it be something like this:
    APQ_DEFINE_COMPOSITE(<namespace>, <schema>, <type-name>, <composite-fields>);
    APQ_DEFINE_CUSTOM_TYPE(<namespace>, <schema>, <type-name>, <type-definition>);
    APQ_DEFINE_ENUM(<namespace>, <schema>, <type-name>, <enum-items>);

Another problem is UTF-8 string representation. Do we need to provide some custom type of validated UTF-8 or just use std::string?